compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -0,0 +1,499 @@
+  AffineExpr dimScale = builder.getAffineConstantExpr(1.0);
+  for (const auto &dimType : order[i]) {
+    switch (dimType) {
+      case DimType::LaneIdX:
+        dim = dim + dimScale * d0;
+        break;
+      case DimType::LaneIdY:
+        dim = dim + dimScale * d1;
+        break;
+      case DimType::LaneIdZ:
+        dim = dim + dimScale * d2;
+        break;
+      default:
+        dim = dim + dimScale * builder.getAffineConstantExpr(state[dimType]);
+        break;
+    }
+    dimScale = dimScale * builder.getAffineConstantExpr(shape[dimType]);
+  }
+  return dim;
+}
+
+std::string Layout::str() const {
+  std::stringstream ss;
+  for (int i = 0; i < DimType::NumDims; i++) {
+    ss << "  " << typeToString(i) << ": [" << shape[i] << "]";
+    bool isRow{false};
+    for (int k = 0; k < order[0].size(); k++) {
+      if (order[0][k] == i) {
+        isRow = true;
+        break;
+      }
+    }
+    if (isRow)
+      ss << "(R)";
+    else
+      ss << "(C)";
+    ss << "  ";
+  }
+  return ss.str();
+}
+
+static MMAType getMMAType(ArrayRef<int64_t> aShape, ArrayRef<int64_t> bShape,
+                          ArrayRef<int64_t> cShape) {
+  if ((aShape[0] % 16 == 0) && (aShape[1] % 16 == 0) && (cShape[0] % 16 == 0) &&
+      (cShape[1] % 8 == 0)) {
+    // B-matrix could be transposed
+    if (((bShape[0] % 16 == 0) && (bShape[1] % 8 == 0)) ||
+        ((bShape[0] % 8 == 0) && (bShape[1] % 16 == 0)))
+      return MMAType::M16N8K16;
+  }
+  return MMAType::NONE;
+}
+
+void setMMALayout(Value aMatrix, Value bMatrix, Value cMatrix,
+                  DenseMap<Value, Layout> &layoutMap) {
+  // First determine which variant of MMA this op is most suitable for
+  auto aType = aMatrix.getType().cast<ShapedType>();
+  auto bType = aMatrix.getType().cast<ShapedType>();
+  auto cType = aMatrix.getType().cast<ShapedType>();
+  ArrayRef<int64_t> aShape = aType.getShape();
+  ArrayRef<int64_t> bShape = bType.getShape();
+  ArrayRef<int64_t> cShape = cType.getShape();
+  MMAType mmaType = getMMAType(aShape, bShape, cShape);
+  if (mmaType == MMAType::NONE) return;
+  // Set layouts for A, B and C
+  auto setLayout = [&](Value matrix, MMAMatrixType matrixType,
+                       bool transpose = false) {
+    DimOrderArray dimOrder;
+    for (int i = 0; i < 2; i++) {
+      dimOrder[i] = getMMADimensions(mmaType, matrixType, i);
+    }
+    std::array<int, 2> canonicalShape =
+        getMMACanonicalShape(mmaType, matrixType);
+    Layout layout(dimOrder, canonicalShape);
+    ArrayRef<int64_t> shape = matrix.getType().cast<ShapedType>().getShape();
+    if (transpose)
+      layout.updateBatchDims(shape[1], shape[0]);
+    else
+      layout.updateBatchDims(shape[0], shape[1]);
+    layoutMap.try_emplace(matrix, layout);
+    // Propagate to preceding transfer_read ops
+    if (auto readOp = matrix.getDefiningOp<vector::TransferReadOp>()) {
+      Value src = readOp.getSource();
+      layoutMap.try_emplace(src, layout);
+    }
```

Comment:
I don't understand that part, the source of transfer_read is a memref, it shouldn't have a layout. Layout should only be attached to vector values.
You shouldn't' need any propagation in this case?

Reply:
Good point! Yes, the memref should not have a layout. I will remove this.